### ISSUE TITLE
SubCellBarCode

### ISSUE THREAD

**Issue by TanerArslan** [Bioconductor/Contributions#1030]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/TanerArslan/SubCellBarCode

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by nturaga** [comment]
Hi @TanerArslan ,

I will review your package only once the build report (http://bioconductor.org/spb_reports/SubCellBarCode_buildreport_20190303094835.html) is clean i.e , no ERROR or WARNING messages and as few as possible NOTE messages. 

Please let me know once you have solved these problems as noted in the build report, saying you are ready for a review.  

Best,

Nitesh

**Comment by TanerArslan** [comment]
Hi @nturaga ,

Thank you for reviewing my package. Currently, i am working on issues and i will let you know when the package is ready for review. 

Best. 

/Taner

**Comment by TanerArslan** [comment]
Hi @nturaga ,
The package was built without errors or warnings. It is ready for the revision. 

Looking forward to hear you soon. 

Best, 
/Taner

**Comment by TanerArslan** [comment]
Hi Nitesh @nturaga , 

The package was ready for the last week. Since i did not hear from you, I hesitated whether i have done everything correct. Should i open a new issue or wait from you? 

Thanks in advance. 

/Taner

**Comment by nturaga** [comment]
I will take a look at it within the next few days. Thanks for the update.

Best,

Nitesh

**Comment by TanerArslan** [comment]
Thanks for the quick respond. 

I am looking forward to hear you soon. 

Cheers. 

/Taner

**Comment by nturaga** [comment]
# SubCellBarCode review

## DESCRIPTION

ok

## NAMESPACE

ok

## man

ok

## data

ok

## vignette

* Title is missing in the vignette.

## R

* In applyThresholdNeighborhood.R, you can just make a helper function
  instead of declaring a function `replacePrediction` within the main
  function. This would be much cleaner, if you make it an private
  function with (`.replacePrediction`). Similar with merge.probability.
  
* In calRowMean.R, the line,

		t.df <- data.frame(Proteins = names(t.df), Fr = unname(t.df))
		
   does not have to be in the lapply loop. Check in other places as
   well to see if you have code which doesn't need to be inside the
   loop (apply statements).
  
  There are some nested loops in probabilityThresholdCompartments.R
  which you can make more efficient very easily. Just move items which
  don't get effected by the loop, outside ot.
  
* This statement can be written more cleanly,

	    if( is.data.frame(protein.data) == FALSE )
			stop('Input must be a data frame format! Type ?loadData')
			
	vs
	
		if (!is.data.frame(protein.data))
			...
			
   Make similar changes everywhere. You don't need to test if
   something is TRUE as well. The if statement only ever enters if TRUE.
   
		if (! is.character(rownames(protein.data)) == TRUE)
			stop('Rownames must be character!')
			
	vs
	
		if (! is.character(rownames(protein.data)))
			...
			
* Unless it's a vector of length > 1, you don't need to use `sprintf`
  for the `message` function. The vector is recycled only if the
  length is greater than 1.
  
		message(sprintf("Number of removed sample-wise proteins: %s",
                    length(sample.removed.prot)))
					
  vs
  
		message("Number of removed sample-wise proteins: ",
                    length(sample.removed.prot))

  Do this everywhere as needed. Make sure you are using an extra
  function like `sprintf` only if it's needed. In most places in your
  code it is not.

	Take a look at : http://bioconductor.org/developers/how-to/coding-style/
	
	
		End-User messages

		message() communicates diagnostic messages (e.g., progress during lengthy computations) during code evaluation.
		warning() communicates unusual situations handled by your code.
		stop() indicates an error condition.
		cat() or print() are used only when displaying an object to the user, e.g., in a show method.

**Comment by TanerArslan** [comment]
Hi Nitesh @nturaga, 

Thanks for the reviewing package. 

I will make the changes as soon as possible. 

/Taner

**Comment by TanerArslan** [comment]
Dear Nitesh @nturaga ,

I have done and made changes about your comment. Below you can see detailed information;

# vignette
* Title is missing in the vignette.

*As you can see the attached code chunk and SubCellBarCode.rmd, the title is not missing on my side and it is present in html output.*

> title: "SubCellBarCode: Integrated workflow for robust classification and 
visualization of spatial proteome"
>  author:
> name: Taner Arslan
  affiliation: "Karolinska Institute, Stockholm, Sweden"
> name: Lukas Orre
  affiliation: "Karolinska Institute, Stockholm, Sweden"
> name: Mattias Vesterlund
  affiliation: "Karolinska Institute, Stockholm, Sweden"
> name: Yanbo Pan
  affiliation: "Karolinska Institute, Stockholm, Sweden"
> name: Janne Lehtiö
  affiliation: "Karolinska Institute, Stockholm, Sweden"
> date: "`r Sys.Date()`"
> output: 
    BiocStyle::html_document:
        toc_fload: true
    BiocStyle::pdf_document: default   
package: SubCellBarCode

*Moreover, you have noticed the same problem in **scMerge** package that you have recently reviewed. I checked its code and it contains title in the rmd file. I wonder whether there is a compatibility issue?*

# R 
* In applyThresholdNeighborhood.R, you can just make a helper function
instead of declaring a function replacePrediction within the main
function. This would be much cleaner, if you make it an private
function with `(.replacePrediction)`. Similar with `merge.probability`.

*I removed all helper functions in applyThresholdNeighborhood.R and defined separate functions. Further, I have noticed that in probabilityThresholdNeighborhood.R possessed same issue and i defined separate functions for it, as well.*

* In calRowMean.R, the line,

> t.df <- data.frame(Proteins = names(t.df), Fr = unname(t.df))

does not have to be in the lapply loop. Check in other places as
well to see if you have code which doesn't need to be inside the
loop (apply statements).

*I removed the `t.df` data frame in the loop.*

* This statement can be written more cleanly, 
> if( is.data.frame(protein.data) == FALSE )
  	stop('Input must be a data frame format! Type ?loadData')

vs 

> if (!is.data.frame(protein.data))
  	...

*I made the changes everywhere.* 

* Unless it's a vector of length > 1, you don't need to use sprintf
for the `message` function. The vector is recycled only if the
length is greater than 1.

> message(sprintf("Number of removed sample-wise proteins: %s",
              length(sample.removed.prot)))

*I removed all redundant `sprintf` in the package.* 

*I looked `end user messages` and replaced for some cases.*  

Looking forward to hear you soon. 

Taner.

**Comment by TanerArslan** [comment]
Dear Nitesh @nturaga, 
I just would like to kindly remind you that package is ready for 17 days for the second round revision. I really want it to be accepted before the new release and very sorry if i am poking you :)

**Comment by nturaga** [comment]
## Update

Hi Taner,

* Depending on how big your matrices can get, checkout the packages
  `matrixStats`, use the functions like rowMeans , colMeans etc to
  make your code more efficient. These functions are highly optimized
  and work in a similar signature to the base R rowMeans as well.

Your package looks mostly ready. It can make the 24th deadline.

Best,

Nitesh

**Comment by TanerArslan** [comment]
Hi @nturaga Nitesh, 

I have added and replaced some with codes with rowsum and rowmean as you suggested. Now the package works well without any problem as expected.
However, I have a slight problem on windows build, i keep getting error on building vignette

>- Invalid Parameter - /figure-html
Warning in shell(paste(c(cmd, args), collapse = " ")) :
  'convert "SubCellBarCode_files/figure-html/coverageMarkers-1.png" -trim "SubCellBarCode_files/figure-html/coverageMarkers-1.png"' execution failed with error code 4

You can also from the build report. I used to get this error before as well. After getting this error, when i resubmitted the package, the problem solved (I literally was not doing anything). Today i tried many times, still i get the problems. Do you know why i keep getting this error?

Thanks in advance. 

/Taner

**Comment by nturaga** [comment]
Hi @TanerArslan 

I think the issue might be a Windows only thing because of a few lines of code. Let's try a couple of things,

1. In this file, https://github.com/TanerArslan/SubCellBarCode/blob/master/R/coveredMarkerProtein.R#L65. Let's change the `plot` to `print(ggplot(....`.

2. It could be the way the path to where these plots end up being a problem. Please take a look at the path, because Linux/Mac do paths in one way, and windows another.

**Comment by TanerArslan** [comment]
Hi @nturaga Nitesh, 

I have tried:
*changing `plot` to `print`*
did not work. 

I just re-pushed, and not it works without error.

**Comment by nturaga** [comment]
Hi @TanerArslan 

I'm not entirely sure what the issue was with the Windows builder. Since it's clean, I'll accept it. But please keep tabs on your package once it has been accepted, specially on the Windows builder.

If it fails again, for the same reason (or similar), bring it to our attention on the bioc-devel mailing list. This would be the maintainer responsibility.

Thanks for your hard work. You'll see some instructions you need to follow once the package is accepted.

Best regards,

Nitesh

**Comment by TanerArslan** [comment]
Dear @nturaga ,

Ok. I will keep track if i encounter again with similar issue. 

Thank you so much for the reviewing and being helpful. 

/Taner

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/TanerArslan.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("SubCellBarCode")`. The package 'landing page' will be created at

https://bioconductor.org/packages/SubCellBarCode

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
